fs-cache/src/cache.rs

```diff
@@ -0,0 +1,668 @@
+        }
+
+        // Sort by modified time (most recent first)
+        file_metadata.sort_by(|a, b| b.2.cmp(&a.2));
+
+        // Clear existing cache
+        self.memory_cache.clear();
+        self.current_memory_bytes = 0;
+
+        // Load files that fit in memory
+        let mut loaded_bytes = 0;
+        let total_bytes: usize = file_metadata
+            .iter()
+            .map(|(_, size, _)| size)
+            .sum();
+
+        for (key, size, _) in file_metadata {
+            if loaded_bytes + size <= self.max_memory_bytes {
+                match self.read_from_disk(&key) {
+                    Ok(value) => {
+                        self.memory_cache
+                            .put(key, CacheEntry { value, size });
+                        loaded_bytes += size;
+                    }
+                    Err(err) => {
+                        log::warn!(
+                            "cache/{}: failed to load key={}: {}",
+                            self.label,
+                            key,
+                            err
+                        );
+                    }
+                }
+            }
+        }
+
+        self.current_memory_bytes = loaded_bytes;
+
+        log::debug!(
+            "cache/{}: loaded {}/{} bytes in memory",
+            self.label,
+            self.current_memory_bytes,
+            total_bytes
+        );
+
+        Ok(())
+    }
+
+    /// Retrieves a value from the memory cache.
+    fn fetch_from_memory(&mut self, key: &K) -> Option<V> {
+        self.memory_cache
+            .get(key)
+            .map(|entry| entry.value.clone())
+    }
+
+    /// Retrieves a value from disk and caches it in memory if possible.
+    fn fetch_from_disk(&mut self, key: &K) -> Option<V> {
+        let file_path = self.path.join(key.to_string());
+        if !file_path.exists() {
+            log::warn!("cache/{}: no value found for key {}", self.label, key);
+            return None;
+        }
+
+        match self.read_from_disk(key) {
+            Ok(value) => {
+                if let Err(err) = self.update_memory_cache(key, &value) {
+                    log::error!(
+                    "cache/{}: failed to add to memory cache for key {}: {}", 
+                    self.label,
+                    key,
+                    err
+                );
+                    return None;
+                }
+                Some(value)
+            }
+            Err(err) => {
+                log::error!(
+                    "cache/{}: failed to load from disk for key {}: {}",
+                    self.label,
+                    key,
+                    err
+                );
+                None
+            }
+        }
+    }
+
+    /// Writes a value to disk using atomic operations.
+    fn persist_to_disk(&mut self, key: &K, value: &V) -> Result<()> {
+        log::debug!("cache/{}: writing to disk for key {}", self.label, key);
+        fs::create_dir_all(&self.path)?;
```

Comment:
I think we should return an error if the cache directory doesn’t exist for any reason. We create the directory when `new` is called (via `load_fs`), so it should always be there. If it’s missing, that’s not a valid state, and we should trigger an error.